plotting.py

```diff
@@ -0,0 +1,253 @@
+    else:
+        plt.xticks(indices, [x.split('.')[1] for x in micro_attack_stages], rotation='vertical')
+    plt.tick_params(axis='x', which='major', labelsize=8)
+    plt.tick_params(axis='x', which='minor', labelsize=8)
+    # plt.yticks(np.arange(0, 13000, 1000))
+    plt.legend([plot[0] for plot in plots], _team_labels)
+    plt.tight_layout()
+    plt.savefig('data_histogram-' + experiment_name + '.png')
+    # plt.show()
+
+
+def plot_episodes(frequencies, episodes, mcat):
+    cap = max(frequencies) + 1
+
+    plt.figure()
+    plt.title(mcat)
+    plt.xlabel('Time ->')
+    plt.ylabel('Slope')
+    plt.plot(frequencies, 'gray')
+    for ep in episodes:
+        xax_start = [ep[0]] * cap
+        xax_end = [ep[1]] * cap
+        yax = list(range(cap))
+
+        plt.plot(xax_start, yax, 'g', linestyle=(0, (5, 10)))
+        plt.plot(xax_end, yax, 'r', linestyle=(0, (5, 10)))
+
+    plt.show()
+    return
+
+
+def _legend_without_duplicate_labels(ax, fontsize=10, loc='upper right'):
+    handles, labels = ax.get_legend_handles_labels()
+    unique = [(h, l) for i, (h, l) in enumerate(zip(handles, labels)) if l not in labels[:i]]
+    unique = sorted(unique, key=lambda x: x[1])
+    ax.legend(*zip(*unique), loc=loc, fontsize=fontsize)
+
+
+def plot_alert_volume_per_episode(tid, attacker_victim, host_episodes, mcats):
+    plt.figure(figsize=(10, 10))
+    ax = plt.gca()
+    plt.title('Micro attack episodes | Team: ' + str(tid) + ' | Host: ' + '->'.join(attacker_victim))
+    plt.xlabel('Time Window (sec)')
+    plt.ylabel('Micro attack stages')
+    # NOTE: Line thicknesses are on per-host basis
+    tmax = max([epi[4] for epi in host_episodes])
+    tmin = min([epi[4] for epi in host_episodes])
+    for idx, ep in enumerate(host_episodes):
+        xax = list(np.arange(ep[0], ep[1] + 1))
+        yax = [mcats.index(ep[2])] * len(xax)
+        thickness = ep[4]
+        lsize = ((thickness - tmin) / (tmax - tmin)) * (5 - 0.5) + 0.5 if (tmax - tmin) != 0.0 else 0.5
+        # lsize = np.log(thickness) + 1 TODO: Either take log or normalize between [0.5 5]
+        msize = (lsize * 2) + 1
+        ax.plot(xax, yax, color=mcols[macro_inv[micro2macro[micro[ep[2]]]]], linewidth=lsize)
+        ax.plot(ep[0], mcats.index(ep[2]), color=mcols[macro_inv[micro2macro[micro[ep[2]]]]], marker='.', linewidth=0,
+                markersize=msize, label=micro2macro[micro[ep[2]]])
+        ax.plot(ep[1], mcats.index(ep[2]), color=mcols[macro_inv[micro2macro[micro[ep[2]]]]], marker='.', linewidth=0,
+                markersize=msize)
+        plt.yticks(range(len(mcats)), [x.split('.')[1] for x in micro.values()], rotation=0)
+    _legend_without_duplicate_labels(ax)
+    plt.grid(True, alpha=0.4)
+
+    # plt.tight_layout()
+    # plt.savefig('Pres-Micro-attack-episodes-Team'+str(tid) +'-Connection'+ attacker[0]+'--'+attacker[1]+'.png')
+    plt.show()
+
+
+def plot_state_groups(state_sequences, data_file):
```

Comment:
Renamed to `make_state_groups` into `plot_state_groups`, since this is more representative of what this method is doing